### PR TITLE
ARROW-4388 [Go] add DimNames() method to tensor Interface

### PR DESCRIPTION
**Pull request by anson627**
https://issues.apache.org/jira/browse/ARROW-4388

### CODE REVIEW

**Comment by anson627** [comment]
@sbinet can you help to take a look?

File: go/arrow/tensor/tensor.go [hunk 1]
```diff
@@ -104,6 +104,7 @@ func (tb *tensorBase) NumDims() int             { return len(tb.shape) }
 func (tb *tensorBase) DimName(i int) string     { return tb.names[i] }
 func (tb *tensorBase) DataType() arrow.DataType { return tb.dtype }
 func (tb *tensorBase) Data() *array.Data        { return tb.data }
+func (tb *tensorBase) DimNames() []string     { return tb.names }
```
Comment (sbinet) [inline go/arrow/tensor/tensor.go]: please `gofmt` your code.
it makes for nice, regular looking, code. (for some definition of "nice". thankfully, all gophers have _now_ the same definition. :P)

a simple:
```
$> cd github.com/apache/arrow/go
$> gofmt -w .
```
should do.

it's a good idea to wire `gofmt` inside your favorite editor (or, even better, use [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) that does pretty much the same thing than `gofmt` but also adds/removes automatically package to/from the `import` block when used/unused.)
see:
- https://github.com/golang/go/wiki/IDEsAndTextEditorPlugins

**Review by sbinet — CHANGES_REQUESTED** [review]
apologies for the belated answer.

you only have added the `DimNames` method to the concrete type (`tensor.tensorBase`.)
you are missing the modification of the actual `tensor.Interface`.

could you please add it?
thanks.

that said, I must admit I am not completely sold on adding `DimNames`...
@stuartcarnie, @alexandreyc  ? what's your take on this?

**Comment by anson627** [comment]
> apologies for the belated answer.
> 
> you only have added the `DimNames` method to the concrete type (`tensor.tensorBase`.)
> you are missing the modification of the actual `tensor.Interface`.
> 
> could you please add it?
> thanks.
> 
> that said, I must admit I am not completely sold on adding `DimNames`...
> @stuartcarnie, @alexandreyc ? what's your take on this?

Thanks for getting back to me! I just added it to the interface.

Just to give you some context, I'm just trying to ramp up on the codebase, by picking up some high level starter task. What's the best way for me to get engaged and be able to contribute better?

File: go/arrow/tensor/tensor.go (hunk 1 above)
Comment (anson627) [reply go/arrow/tensor/tensor.go]: > please `gofmt` your code.
> it makes for nice, regular looking, code. (for some definition of "nice". thankfully, all gophers have _now_ the same definition. :P)
> 
> a simple:
> 
> ```
> $> cd github.com/apache/arrow/go
> $> gofmt -w .
> ```
> should do.
> 
> it's a good idea to wire `gofmt` inside your favorite editor (or, even better, use [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) that does pretty much the same thing than `gofmt` but also adds/removes automatically package to/from the `import` block when used/unused.)
> see:
> 
> * https://github.com/golang/go/wiki/IDEsAndTextEditorPlugins

Thanks for the suggestion! I've configured my GoLand File Watcher to do "go fmt".

**Comment by sbinet** [comment]
@stuartcarnie, @alexandreyc: ping?

**Comment by sbinet** [comment]
@anson627 sending PRs is the best way :)
a possibly less "contentious" PR would be perhaps https://issues.apache.org/jira/browse/ARROW-3951 (it's just implementing a CSV writer, using the CSV reader as a symetrical blueprint.)

**Comment by anson627** [comment]
> @anson627 sending PRs is the best way :)
> a possibly less "contentious" PR would be perhaps https://issues.apache.org/jira/browse/ARROW-3951 (it's just implementing a CSV writer, using the CSV reader as a symetrical blueprint.)

Sure, I will take a look at this task. Thanks!

**Comment by anson627** [comment]
> @stuartcarnie, @alexandreyc: ping?

Thank! I will try to follow up with them! I double check the build, the failure was unrelated.

**Comment by stuartcarnie** [comment]
@sbinet I am not familiar with `tensor.Interface` usage enough to say whether this API is necessary or useful. @anson627 do you have a use case for this API?

**Comment by sbinet** [comment]
I think @rcoreilly is the one who originally filed the issue on JIRA. He may want to comment :)
(I suspect that's for his work on github.com/emer/emergent/dtable)

**Comment by rcoreilly** [comment]
yep DimNames() just makes it much easier to convert between arrow tensor and other implementations of tensor -- otherwise you would have to manually reconstruct the DimNames slice by iterating over the DimName method.  Not the biggest thing in the world but definitely seems worth including given how trivial it is to do so.  And adding it consistently to the interface as well as the impl makes sense.  Thanks!

**Comment by anson627** [comment]
> yep DimNames() just makes it much easier to convert between arrow tensor and other implementations of tensor -- otherwise you would have to manually reconstruct the DimNames slice by iterating over the DimName method. Not the biggest thing in the world but definitely seems worth including given how trivial it is to do so. And adding it consistently to the interface as well as the impl makes sense. Thanks!

Thanks @rcoreilly for filling in the context!

**Comment by anson627** [comment]
@stuartcarnie Do you any other further question? How can I merge this code change?

**Comment by stuartcarnie** [comment]
@anson627 no, looks good to me 👍

**Review by sbinet — APPROVED** [review]

**Comment by anson627** [comment]
@sbinet how can we merge and close this change?

**Comment by sbinet** [comment]
if nobody (or @wesm ) beats me to it, I'll merge this tomorrow (when in front of a real keyboard)

**Pull request closed without merging**